c7n/resources/iam.py

```diff
@@ -1953,11 +1953,32 @@ class UserPolicy(ValueFilter):
               - type: policy
                 key: PolicyName
                 value: AdministratorAccess
+                include-via: true
     """
 
-    schema = type_schema('policy', rinherit=ValueFilter.schema)
+    schema = type_schema('policy', rinherit=ValueFilter.schema, 
+        **{'include-via': {'type': 'boolean'}})
     schema_alias = False
-    permissions = ('iam:ListAttachedUserPolicies',)
+    permissions = (
+        'iam:ListAttachedUserPolicies',
+        'iam:ListGroupsForUser',
+        'iam:ListAttachedGroupPolicies',
+    )
+
+    def user_groups_policies(self, client, u):
+        if 'c7n:GroupsPolicies' not in u:
+            u['c7n:GroupsPolicies'] = []
+
+        u['c7n:Groups'] = client.list_groups_for_user(
+            UserName=u['UserName'])['Groups']
+        for ug in u['c7n:Groups']:
+            aps = client.list_attached_group_policies(
+                GroupName=ug['GroupName'])['AttachedPolicies']
+            for ap in aps:
+                u['c7n:GroupsPolicies'].append(
+                    client.get_policy(PolicyArn=ap['PolicyArn'])['Policy'])
```

Comment:
this feels like we're potentially fetching the policies redundantly, ideally we get the set of policies arns for a set of users, and only fetch a given policy once. for common group memberships, this will significantly reduce api calls, ditto for policies attached to multiple groups.

Reply:
I modified the code to avoid redundant api calls. Please review it again.

Reply:
afaics the changes doesn't seem like it address the issue.

ie say I have 5 users, 2 groups, and 1 policy attached to each group.

with current implementation this will fetch the policy at least 5 times, since its operating and looking up per user. it has to operate on the user set level to do the api call minimization.

user set -> group set -> policy set. with each of the transitions once is the goal. to do that you need an up front initialization of data structures for the indirect via option while looking at the whole set of users to minimize the redundant api calls on looking up a policy for a group thats already been looked up.

Reply:
I got it. I updated code and test case.

- test
  - case
    - user 1
      - group : group 1
      - user policy : a
    - user 2
      - group : group 1
      - user policy :
    - group 1
      - policy : a, b
  - api call count
    - iam.ListUsers : 1
    - iam.GetUser : 2
    - iam.ListAttachedUserPolicies : 2
    - iam.ListGroupsForUser : 2
    - iam.ListAttachedGroupPolicies : 1 
    - iam.GetPolicy : 2